image/manifest.go

```diff
@@ -26,28 +26,274 @@ import (
+func (sm *SeedManifest) MarkSnapRevisionSeeded(snapName string, revision int) error {
+	snapRev := snap.R(revision)
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		// Allowed revision specified, it must match.
+		if rev.Revision != snapRev {
+			return fmt.Errorf("revision %s does not match the allowed revision %s", snapRev, rev.Revision)
+		}
+	}
+
+	if rev, ok := sm.revsSeeded[snapName]; ok {
+		// Already marked as seeding.
+		return fmt.Errorf("cannot mark %q (%s) as seeded, it has already been marked seeded: %q (%s)",
+			snapName, snapRev, rev.SnapName, rev.Revision)
+	}
+
+	sm.revsSeeded[snapName] = &SeedManifestSnapRevision{
+		SnapName: snapName,
+		Revision: snapRev,
+	}
+	return nil
+}
+
+// MarkValidationSetSeeded marks a validation-set as seeded. It verifies against any previously
+// set rules by SetAllowedValidationSet, and sets up new rules based on the snaps defined in the
+// validation set.
+// This relies on validation-set assertions being marked here the moment they are fetched by
+// the seedwriter, which should be done before the first call to MarkSnapRevisionSeeded.
+func (sm *SeedManifest) MarkValidationSetSeeded(vsa *asserts.ValidationSet, pinned bool) error {
+	vs := newSeedManifestValidationSet(vsa, pinned)
+	if _, ok := sm.vsSeeded[vs.Unique()]; ok {
+		return fmt.Errorf("cannot mark validation set %q as seeded, it has already been marked as such",
+			vs.Unique())
+	}
+
+	// Check against any pre-defined restrictions.
+	if allowed, ok := sm.vsAllowed[vs.Unique()]; ok {
+		if allowed.Sequence != vs.Sequence {
+			return fmt.Errorf("sequence of %q (%d) does not match the allowed sequence (%d)",
+				vs.Unique(), vs.Sequence, allowed.Sequence)
+		}
+		if allowed.Pinned != pinned {
+			return fmt.Errorf("pinning of %q (%t) does not match the allowed pinning (%t)",
+				vs.Unique(), pinned, allowed.Pinned)
+		}
+	}
+
+	for _, sn := range vsa.Snaps() {
+		if sn.Presence == asserts.PresenceInvalid {
+			continue
+		}
+
+		// Update allowed snaps based on the validation-set.
+		if err := sm.SetAllowedSnapRevision(sn.SnapName(), sn.Revision); err != nil {
+			return err
+		}
+
+		// For book-keeping purposes we add the snap to the list of controlled
+		// snap revisions by the validation set.
+		vs.Snaps = append(vs.Snaps, sn.SnapName())
+	}
+
+	sm.vsSeeded[vs.Unique()] = vs
+	return nil
+}
+
+// AllowedRevision retrieves any specified revision rule for the snap
+// name.
+func (sm *SeedManifest) AllowedSnapRevision(snapName string) snap.Revision {
+	// TODO: Check seeded validation-sets as well.
+	if rev, ok := sm.revsAllowed[snapName]; ok {
+		return rev.Revision
+	}
+	return snap.Revision{}
+}
+
+// ValidationSetsAllowed returns the validation sets specified as allowed.
+func (sm *SeedManifest) AllowedValidationSets() []*SeedManifestValidationSet {
+	var vss []*SeedManifestValidationSet
+	for _, vs := range sm.vsAllowed {
+		vss = append(vss, vs)
+	}
+	return vss
```

Comment:
the sorting of the result might vary because of the map, tests need to be careful about that, or we need to sort here

Reply:
Yea nice catch. It would definitely have broken tests in the long run